### PR TITLE
Add parser tests from nim-regex

### PR DESCRIPTION
**Pull request by danmoseley**
Contributes to https://github.com/dotnet/runtime/issues/61895

What we really care about is the functional tests, but getting this part committed separately.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 1]
```diff
@@ -709,23 +709,34 @@ public void ParseCheckOffset(string pattern, RegexOptions options, RegexParseErr
             Parse(pattern, options, error, offset);
         }
 
+
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:712]: Nit: why the extra blank line?

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 2]
```diff
@@ -709,23 +709,34 @@ public void ParseCheckOffset(string pattern, RegexOptions options, RegexParseErr
...
         }
 
+        private static void LogActual(string pattern, RegexOptions options, RegexParseError error, int offset)
+        {         
+            string s = (error == RegexParseError.Unknown) ?
+                @$"        [InlineData(@""{pattern}"", RegexOptions.{options.ToString()}, null)]" :
+                @$"        [InlineData(@""{pattern}"", RegexOptions.{options.ToString()}, RegexParseError.{error.ToString()}, {offset})]";
+
+            // File.AppendAllText(@"/tmp/out.cs", s + "\n"); // for updating baseline
+        }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:738]: I don't understand... what's the purpose of this method?

**Review by stephentoub — COMMENTED** [review]
Other than the CI failures from a couple of the new tests, LGTM.

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:738]: It updates the baseline: I can take a bunch of presumably-invalid regex inputs from someplace, and turn them into this format:
```
[InlineData("SOMEREGEX1", RegexOptions.None, null)]
[InlineData("SOMEREGEX2", RegexOptions.None, null)]
...
```

I paste those into the test, uncomment this line, and run the test, and it emits into that file exactly what I need to paste in  eg
```
[InlineData(@"SOMEREGEX1", RegexOptions.None, RegexParseError.UnrecognizedEscape, 3)]
[InlineData(@"SOMEREGEX2", RegexOptions.None, InsufficientClosingParentheses, 2)]
...
```
I then replace what I pasted in with those, and I have my updated test.

Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:738]: is that OK? I'm finding this a useful approach: have the test construct the baseline. 

Of course, I'll eyeball to satisfy myself that the results are good. In some cases, I'm emitting in a different format locally to diff against the original test output as well.

Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:749]: Added a comment.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:749]: Is it possible to do easily as a standalone, commented out test?  I think what I'm objecting to is it's modifying other tests with dead code that is almost certain to rot.

**Review by joperezr — APPROVED** [review]
LGTM, Thanks Dan.

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs:749]: Sure, I can change it to that.

**Comment by danmoseley** [comment]
I'll do that cleanup in my next PR.

**Pull request merged**